domain/port/watcher_test.go

```diff
@@ -0,0 +1,296 @@
+		err := s.srv.UpdateUnitPorts(context.Background(), s.unitUUIDs[0], network.GroupedPortRanges{
+			"ep0": {ssh},
+		}, network.GroupedPortRanges{})
+		c.Assert(err, jc.ErrorIsNil)
+	}, func(w watchertest.WatcherC[[]string]) {
+		w.Check(watchertest.StringSliceAssert(machineUUIDs[0]))
+	})
+
+	// open a port on an endpoint with opened ports on a unit on machine 0
+	harness.AddTest(func(c *gc.C) {
+		err := s.srv.UpdateUnitPorts(context.Background(), s.unitUUIDs[0], network.GroupedPortRanges{
+			"ep0": {http},
+		}, network.GroupedPortRanges{})
+		c.Assert(err, jc.ErrorIsNil)
+	}, func(w watchertest.WatcherC[[]string]) {
+		w.Check(watchertest.StringSliceAssert(machineUUIDs[0]))
+	})
+
+	// open a port on a new endpoint on another unit on machine 0
+	harness.AddTest(func(c *gc.C) {
+		err := s.srv.UpdateUnitPorts(context.Background(), s.unitUUIDs[1], network.GroupedPortRanges{
+			"ep1": {http},
+		}, network.GroupedPortRanges{})
+		c.Assert(err, jc.ErrorIsNil)
+	}, func(w watchertest.WatcherC[[]string]) {
+		w.Check(watchertest.StringSliceAssert(machineUUIDs[0]))
+	})
+
+	// open a port on a endpoint on a unit on machine 1
+	harness.AddTest(func(c *gc.C) {
+		err := s.srv.UpdateUnitPorts(context.Background(), s.unitUUIDs[2], network.GroupedPortRanges{
+			"ep2": {https},
+		}, network.GroupedPortRanges{})
+		c.Assert(err, jc.ErrorIsNil)
+	}, func(w watchertest.WatcherC[[]string]) {
+		w.Check(watchertest.StringSliceAssert(machineUUIDs[1]))
+	})
+
+	// open a port that's already open
+	harness.AddTest(func(c *gc.C) {
+		err := s.srv.UpdateUnitPorts(context.Background(), s.unitUUIDs[0], network.GroupedPortRanges{
+			"ep0": {ssh},
+		}, network.GroupedPortRanges{})
+		c.Assert(err, jc.ErrorIsNil)
+	}, func(w watchertest.WatcherC[[]string]) {
+		w.AssertNoChange()
+	})
+
+	// close a port on an endpoint on a unit on machine 0
+	harness.AddTest(func(c *gc.C) {
+		err := s.srv.UpdateUnitPorts(context.Background(), s.unitUUIDs[0], network.GroupedPortRanges{}, network.GroupedPortRanges{
+			"ep0": {ssh},
+		})
+		c.Assert(err, jc.ErrorIsNil)
+	}, func(w watchertest.WatcherC[[]string]) {
+		w.Check(watchertest.StringSliceAssert(machineUUIDs[0]))
+	})
+
+	// close the final open port of an endpoint for a unit on machine 0
+	harness.AddTest(func(c *gc.C) {
+		err := s.srv.UpdateUnitPorts(context.Background(), s.unitUUIDs[0], network.GroupedPortRanges{}, network.GroupedPortRanges{
+			"ep0": {http},
+		})
+		c.Assert(err, jc.ErrorIsNil)
+	}, func(w watchertest.WatcherC[[]string]) {
+		w.Check(watchertest.StringSliceAssert(machineUUIDs[0]))
+	})
+
+	// close a port range which isn't open
+	harness.AddTest(func(c *gc.C) {
+		err := s.srv.UpdateUnitPorts(context.Background(), s.unitUUIDs[1], network.GroupedPortRanges{}, network.GroupedPortRanges{
+			"ep1": {https},
+		})
+		c.Assert(err, jc.ErrorIsNil)
+	}, func(w watchertest.WatcherC[[]string]) {
+		w.AssertNoChange()
+	})
+
+	harness.Run(c, []string{machineUUIDs[0], machineUUIDs[1]})
```

Comment:
**question:** Is there a use case where an update cause several machine to be triggered ? If it is, there is a missing test case.

Reply:
Potentially, if we have two changes in quick succession, the events could be combined

I'll add some tests for this